Readme.md
# plex-api [![Build Status](https://api.travis-ci.org/phillipj/node-plex-api.png?branch=master)](http://travis-ci.org/phillipj/node-plex-api)

Small module which helps you query the Plex Media Server HTTP API.

## Usage

```bash
$ npm install plex-api --save
```

**PlexAPI(options | hostname)**

Instantiate a PlexAPI client.

The parameter can be a string representing the server's hostname, or an object with the following properties:

Options:
- **hostname**: hostname where Plex Server runs
- **port**: port number Plex Server is listening on (optional, default: `32400`)
- **https**: (optional, default: `false`)
- **username**: plex.tv username (optional / required for PlexHome)
- **password**: plex.tv password (optional / required for PlexHome)
- **managedUser**: details required to perform operations as a managed PlexHome user
	- **name**: managed user name
	- **pin**: optional pin code for the managed user
- **token**: plex.tv authentication token (optional)
- **timeout**: timeout value in milliseconds to use when making requests (optional)
- **responseParser**: custom function to be used parsing all responses from Plex Server (optional)
- **requestOptions**: add or replace options to be provided to [request.js](https://www.npmjs.com/package/request) when performing API request (optional)
- **options**: override additional PlexHome options (optional, but recommended for PlexHome)
	- **identifier**: A unique client identifier. Default is a `generated uuid v4`. *Note: you should really provide this rather than let it get generated. Every time your app runs, a new "device" will get registered on your Plex account, which can lead to poor performance once hundreds or thousands of them get created. Trust me!*
	- **product**: The name of your application. Official Plex examples: `Plex Web`, `Plex Home Theater`, `Plex for Xbox One`. Default `Node.js App`
	- **version**: The version of your app. Default `1.0`
	- **deviceName**: The "name" of the device your app is running on. For apps like Plex Home Theater and mobile apps, it's the computer or phone's name chosen by the user. Default `Node.js App`
	- **platform**: The platform your app is running on. The use of this is inconsistent in the official Plex apps. It is not displayed on the web interface. Official Plex examples: `Chrome`, `Plex Home Theater`, `Windows`. Default is `Node.js`.
	- **platformVersion**: The platform version. Default is the version of Node running.
	- **device**: The name of the type of computer your app is running on, usually the OS name. Official Plex examples: `Windows`, `iPhone`, `Xbox One`. Default is whatever `os.platform()` returns.

Here's an example of what an app shows up as on the Plex web interface

![Plex Device Example](docs/plex-device-example.png?raw)

The rows in that example from top to bottom are `deviceName`, `version`, `product`, and `device`.

### .query(options)

**Retrieve content from URI**

The parameter can be a string representing the URI, or an object with the following properties:
- **uri**: the URI to query
- (optional) **extraHeaders**: an object with extra headers to send in the HTTP request. Useful for things like X-Plex-Target-Client-Identifier

Aside from requesting the API and returning its response, an `.uri` property are created to easier follow the URIs available in the HTTP API. At the moment URIs are attached for Directory and Server items.

```js
var PlexAPI = require("plex-api");
var client = new PlexAPI("192.168.0.1");

client.query("/").then(function (result) {
	console.log("%s running Plex Media Server v%s",
		result.friendlyName,
		result.version);

	// array of children, such as Directory or Server items
	// will have the .uri-property attached
	console.log(result._children);
}, function (err) {
	console.error("Could not connect to server", err);
});
```

### .postQuery(options)

**Send a POST request and retrieve the response**

This is identical to `query()`, except that the request will be a POST rather than a GET. It has the same required and optional parameters as `query()`.

Note that the parameters can only be passed as a query string as part of the uri, which is all Plex requires. (`Content-Length` will always be zero)

```js
var PlexAPI = require("plex-api");
var client = new PlexAPI("192.168.0.1");

client.postQuery("/playQueue?type=video&uri=someuri&shuffle=0").then(function (result) {
	console.log("Added video to playQueue %s",
		result.playQueueID);

	// array of children, such as Directory or Server items
	// will have the .uri-property attached
	console.log(result._children);
}, function (err) {
	console.error("Could not connect to server", err);
});
```

### .putQuery(options)

**Send a PUT request and retrieve the response**

This is identical to `query()`, except that the request will be a PUT rather than a GET. It has the same required and optional parameters as `query()`. It's is used to update parts of your Plex library.

Note that the parameters can only be passed as a query string as part of the uri, which is all Plex requires. (`Content-Length` will always be zero)

```js
var PlexAPI = require("plex-api");
var client = new PlexAPI("192.168.0.1");

client.putQuery("/library/sections/3/all?type=1&id=123&summary.value=updatedSummaryText")
	.then(function (result) {
		console.log("Description of video by id 123 has been set to 'updatedSummaryText'");
	}, function (err) {
		console.error("Could not connect to server", err);
	});
```

### .deleteQuery(options)

**Send a DELETE request and retrieve the response**

This is identical to `query()`, except that the request will be a DELETE rather than a GET. It has the same required and optional parameters as `query()`. It's is used to delete parts of your Plex library.

Note this will also delete the media files on hard disk! This can be allowed or forbidden in the plex-media-server options.
Returns status code 403 if delete is not allowed in the plex-media-server options.

```js
var PlexAPI = require("plex-api");
var client = new PlexAPI("192.168.0.1");

client.deleteQuery("/library/metadata/10001/media/2002")
	.then(function () {
		console.log("Media was successfully deleted");
	}, function (err) {
		console.error("Could not connect to server", err);
	});
```

### .perform(options)

**Perform an API action**

When performing an "action" on the HTTP API, the response body will be empty.
As the response content itself is worthless, `perform()` acts on the HTTP status codes the server responds with.
It has the same required and optional parameters as `query()`.

```js
var PlexAPI = require("plex-api");
var client = new PlexAPI("192.168.0.1");

// update library section of key "1"
client.perform("/library/sections/1/refresh").then(function () {
	// successfully started to refresh library section #1
}, function (err) {
	console.error("Could not connect to server", err);
});
```

### .find(options, [{criterias}])

**Find matching child items on URI**

Uses `query()` behind the scenes, giving all directories and servers the beloved `.uri` property. It has the same required and optional parameters as `query`, in addition to a second optional `criterias` parameter.

```js
var PlexAPI = require("plex-api");
var client = new PlexAPI("192.168.0.1");

// filter directories on Directory attributes
client.find("/library/sections", {type: "movie"}).then(function (directories) {
	// directories would be an array of sections whose type are "movie"
}, function (err) {
	console.error("Could not connect to server", err);
});

// criterias are interpreted as regular expressions
client.find("/library/sections", {type: "movie|shows"}).then(function (directories) {
	// directories type would be "movie" OR "shows"
}, function (err) {
	console.error("Could not connect to server", err);
});

// shorthand to retrieve all Directories
client.find("/").then(function (directories) {
	// directories would be an array of Directory items
}, function (err) {
	throw new Error("Could not connect to server");
});
```

## Authenticators

An authenticator is used by plex-api to authenticate its request against Plex Servers with a PlexHome setup. The most common authentication mechanism is by username and password.

You can provide your own custom authentication mechanism, read more about custom authenticators below.

### Credentials: username and password

Comes bundled with plex-api. Just provide `options.username` and `options.password` when creating a PlexAPI instance and you are good to go.

See the [plex-api-credentials](https://www.npmjs.com/package/plex-api-credentials) module for more information about its inner workings.

### PIN: authenticate by PIN code

An authentication module that provides an interface for authenticating with Plex using a PIN, like the official clients do.

https://www.npmjs.com/package/plex-api-pinauth

### Custom authenticator

In its simplest form an `authenticator` is an object with **one required** function `authenticate()` which should return the autentication token needed by plex-api to satisfy Plex Server.

An optional method `initialize()` could be implemented if you need reference to the created PlexAPI instance when it's created.

```js
{
  // OPTIONAL
  initialize: function(plexApi) {
    // plexApi === the PlexAPI instance just created
  },
  // REQUIRED
  authenticate: function(plexApi, callback) {
    // plexApi === the PlexAPI instance requesting the authentication token

    // invoke callback if something fails
    if (somethingFailed) {
      return callback(new Error('I haz no cluez about token!'));
    }

    // or when you have a token
    callback(null, 'I-found-this-token');
  }
}
```

## Response parsing

You can provide a custom function responsible for parsing all Plex Server responses if you need to, by providing it in the `responseParser` when instantiating a Plex API client.

The default implementation either parses the JSON in the response from the Plex Server, converts XML to a JavaScript object or returns the response as is, depending on the response Content-Type header.

A response parsing function gets two arguments provided: `response`, `body` and is expected to return a `Promise`.

```js
function myCustomJsonResponseParser(response, body) {
	const bodyAsString = body.toString('ut8');

	return Promise.resolve(bodyAsString).then(JSON.parse);
}
```

## HTTP API Documentation
For more information about the API capabilities, see the [unofficial Plex API documentation](https://github.com/Arcanemagus/plex-api/wiki). The [PlexInc's desktop client wiki](https://github.com/plexinc/plex-media-player/wiki/Remote-control-API) might also be valueable.

## Running tests

```shell
$ npm install
$ npm test
```

Automatically run all tests whenever files has been changed:
```shell
$ npm run test:watch
```

**NB!** This project runs [prettier](https://github.com/prettier/prettier) for code formatting.
The test script will fail if prettier rules are not followed. Scripts can be formatted by running:

```bash
$ npm run format
```

or by adding an editor plugin/extension as described in the [prettier](https://github.com/prettier/prettier) project.

## Usage in the wild

plex-api has proven to be useful in more than one project over the years.

Do you have project which uses plex-api? Please tell us about it and we'll list it here :)

### alexa-plex

Alexa (Amazon Echo) app for interacting with a Plex Server and controlling client playback.

https://github.com/OverloadUT/alexa-plex by [@OverloadUT](https://github.com/OverloadUT).

### nl.kikkert.plex

The Plex Remote control app for the Homey device.

https://github.com/MikeOne/nl.kikkert.plex by [@MikeOne](https://github.com/MikeOne).

### plex2netflix

See how much of your media from Plex is available on Netflix with a CLI command.

https://github.com/SpaceK33z/plex2netflix by [@SpaceK33z](https://github.com/SpaceK33z).

## Contributing

Contributions are more than welcome! Create an issue describing what you want to do. If that feature is seen to fit this project, send a pull request with the changes accompanied by tests.

## License
(The MIT License)

Copyright (c) 2013-2018 Phillip Johnsen &lt;johphi@gmail.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
## Change log

### v5.3.2

#### Fixed

- Fix `.find()` query selecting wrong children object by [@hkjasf687261gh](https://github.com/hkjasf687261gh) [#114](https://github.com/phillipj/node-plex-api/pull/114)

### v5.3.1

#### Fixed

- If Content-Type is missing, default to a UTF-8 string response by [@nicjansma](https://github.com/nicjansma) [#105](https://github.com/phillipj/node-plex-api/pull/105)

### v5.3.0

#### Added

- Allow additional options for request by [@Zefau](https://github.com/Zefau) [#104](https://github.com/phillipj/node-plex-api/pull/104)

### v5.2.5

#### Changed
- Updated mocha because of vulnerabilities by [@phillipj](https://github.com/phillipj)

### v5.2.4

#### Changed
- Updated plex-api-credentials & lodash of child dependencies because of vulnerabilities by [@phillipj](https://github.com/phillipj)

### v5.2.3

#### Fixed
- Basic support for handling responses from Plex Media Server v1.3 (and above) by [@Andy1210](https://github.com/Andy1210) [#92](https://github.com/phillipj/node-plex-api/pull/92)

### v5.2.2

#### Changed
- Updated lodash v4.17.10 -> v4.17.11 by [@phillipj](https://github.com/phillipj) [#100](https://github.com/phillipj/node-plex-api/pull/100)

### v5.2.1

#### Changed
- Updated request v2.79.0 -> v2.87.0 by [@au5ton](https://github.com/au5ton) [#95](https://github.com/phillipj/node-plex-api/pull/95)

### v5.2.0

#### Added
- Add .deleteQuery() by [@Mario-F](https://github.com/Mario-F) [#89](https://github.com/phillipj/node-plex-api/pull/89)

### v5.1.0

#### Added
- Allow custom response parser to be provided upon instantiation by @phillipj [76539c](https://github.com/phillipj/node-plex-api/commit/76539c8bca455607d3f2aced0b20d91a13d1b936)

### v5.0.2

#### Fixed
- reject() instead of throw when authenticating by @dabbers [#82](https://github.com/phillipj/node-plex-api/pull/82)

### v5.0.1

#### Changed
- Updated request v2.72.0 -> 2.79.0 by @SpaceK33z [#78](https://github.com/phillipj/node-plex-api/pull/78)

### v5.0.0

#### Changed
- Replace use of Q with native Promise by @phillipj [#77](https://github.com/phillipj/node-plex-api/pull/77)

### Fixed
- Fix undefined runtime error when authenticating by @phillipj [#76](https://github.com/phillipj/node-plex-api/pull/76)

#### BREAKING CHANGE

All methods now return native Promise instances, rather than Q promises as before.

### v4.0.0

#### Added
- Support for managed users by @hyperlink [#70](https://github.com/phillipj/node-plex-api/pull/70) <br>
  By specifying `options.managedUser` when creating a plex-api client, see more in [Readme.md](./Readme.md).

  [About managed users on support.plex.tv](https://support.plex.tv/hc/en-us/articles/203948776-Managed-Users).

#### BREAKING CHANGE

Requires at least Node.js v4.0.

### v3.5.0

#### Added
- Added support for PUT queries with .putQuery() by @IonicaBizau [#61](https://github.com/phillipj/node-plex-api/pull/61)

### v3.4.0

#### Added
- Added timeout option to control timeout on subsequent requests by @lokenx [#60](https://github.com/phillipj/node-plex-api/pull/60)

#### Updates
- xml2js from v0.4.15 to v0.4.16
- request from v2.67.0 to v2.69.0

### v3.3.0

#### Added
- Add support for extra headers in all of the query() related methods by @OverloadUT [#48](https://github.com/phillipj/node-plex-api/pull/48)
- Add support for `https` parameter to force https even on non-443 port by @OverloadUT [#47](https://github.com/phillipj/node-plex-api/pull/47)

#### Fixed
- Enabling gzip to fix a bug in some versions of PMS/PHT that silently fail when no Accept-Encoding header is sent by @OverloadUT [#51](https://github.com/phillipj/node-plex-api/pull/51)

### v3.2.0
- Added `options.token` to specify authentication token at PlexAPI client instantiation by @MikeOne
- Made responses with status code 2xx considered successfull, not just 200 by @MikeOne

### v3.1.2
- Fixed XML parsing by @phillipj

### v3.1.1
- Fixing broken authentication blooper by @phillipj

### v3.1.0

- Extensible authentication mechanisms by @phillipj
- Extracted HTTP headers generation into [plex-api-headers](https://www.npmjs.com/package/plex-api-headers) for re-use by @phillipj

### v3.0.0
- Change xml2json to xml2js by @OverloadUT

#### BREAKING CHANGE

Some of URIs on the Plex Server responds with XML instead of JSON. Previous versions of plex-api used xml2json to translate between XML to JSON. We have now replaced xml2json with xml2js which might result in a different JSON format when requesting URIs responding with XML.

Please see the documentation of [xml2json](https://github.com/buglabs/node-xml2json) and [xml2js](https://github.com/Leonidas-from-XIV/node-xml2js) for more details about their differences.

### v2.5.0
- Updated the readme to explain each of the X-Plex headers by @OverloadUT
- The X-Plex headers are now sent on every request by @OverloadUT
- Added missing X-Plex headers: deviceName, platformVersion by @OverloadUT
- Updated the default X-Plex headers to be a bit more descriptive by @OverloadUT

### v2.4.0
- Added `postQuery()` to perform POST requests by @OverloadUT

### v2.3.0
- PlexHome authentication if needed when calling `.perform()` as with `.query()`, by @OverloadUT

### v2.2.0
- Convert to JSON or XML according to server response header, or resolve with raw server response buffer. This allows for image buffers to be fetched. By @YouriT

### v2.1.0
- Add ability to define app options by @DMarby

### v2.0.1
- Bugfix for wrong `.uri` in some cases by @pjeby

### v2.0.0
- PlexHome support
- Deprecated port argument of PlexAPI constructor in favor of an options object
- Retrieves JSON from the Plex HTTP API instead of XML **see breaking changes below!**

#### BREAKING CHANGES FROM v1.0.0 AND BELOW

We're now retrieving JSON from the Plex HTTP API instead of XML, which got translated to JSON by this module. Direct consequences:
- Attributes previously found in `result.attributes` are now available directly in `result`

test/buster.js
var config = module.exports;

config["Plex API tests"] = {
	env: "node",
	rootPath: "../",
	sources: [
		"lib/api.js"
	],
	tests: [
		"test/*-test.js"
	]
};

test/server.js
var fs = require('fs');
var nock = require('nock');

var PLEX_SERVER_PORT = 32400;

var respondWith;

function hasExtension(filename) {
    return filename.indexOf('.') !== -1;
}

function replaceSlashWithRoot(uri) {
    return uri.replace(/^\/$/, '/root');
}

function respondToRequest(uri, requestBody, cb) {
    uri = replaceSlashWithRoot(uri);

    var filepath = hasExtension(uri) ? uri : uri + '.json';
    if (respondWith === 'content') {
        fs.readFile('test/samples/' + filepath, cb);
    } else if (respondWith === 'failure') {
        return cb(new Error('Server decided to fail...'));
    } else {
        cb(null);
    }
}

// Looks kinda strange, but its needed for nock
// not to explode as we've got one .get('/') in our nock scope
function replaceActualPathToRoot(path) {
    return '/';
}

module.exports = {
    start: function start(options) {
        options = options || {};
        options.schemeAndHost = options.schemeAndHost || 'http://localhost';
        options.port = options.port || PLEX_SERVER_PORT;
        options.contentType = options.contentType || 'application/json';
        respondWith = 'content';

        var scope = nock(options.schemeAndHost + ':' + options.port, {
            reqheaders: options.reqheaders,
        })
            .defaultReplyHeaders({
                'Content-Type': options.contentType,
            })
            .filteringPath(replaceActualPathToRoot)
            .get('/')
            .reply(options.statusCode || 200, respondToRequest);

        // NOT TO PLEASED ABOUT HARDCODING THIS MATCHHEADER() TOKEN ...
        if (options.expectRetry) {
            scope
                .get('/')
                .matchHeader('X-Plex-Token', 'abc-pretend-to-be-token')
                .reply(options.retryStatusCode || 200, respondToRequest);
        }

        return scope;
    },

    empty: function empty() {
        return nock('http://localhost' + ':' + PLEX_SERVER_PORT);
    },

    expectsPost: function expectsPost(options) {
        options = options || {};
        options.port = options.port || PLEX_SERVER_PORT;
        options.contentType = options.contentType || 'application/json';
        respondWith = 'content';

        return nock('http://localhost:' + options.port, {
            reqheaders: options.reqheaders,
        })
            .defaultReplyHeaders({
                'Content-Type': options.contentType,
            })
            .filteringPath(replaceActualPathToRoot)
            .post('/')
            .reply(options.statusCode || 200, respondToRequest);
    },

    expectsPut: function expectsPut(options) {
        options = options || {};
        options.port = options.port || PLEX_SERVER_PORT;
        options.contentType = options.contentType || 'application/json';
        respondWith = 'content';

        return nock('http://localhost:' + options.port, {
            reqheaders: options.reqheaders,
        })
            .defaultReplyHeaders({
                'Content-Type': options.contentType,
            })
            .filteringPath(replaceActualPathToRoot)
            .put('/')
            .reply(options.statusCode || 200, respondToRequest);
    },

    expectsDelete: function expectsDelete(options) {
        options = options || {};
        options.port = options.port || PLEX_SERVER_PORT;
        options.contentType = options.contentType || 'text/html';
        respondWith = null;

        return nock('http://localhost:' + options.port, {
            reqheaders: options.reqheaders,
        })
            .defaultReplyHeaders({
                'Content-Type': options.contentType,
            })
            .filteringPath(replaceActualPathToRoot)
            .delete('/')
            .reply(options.statusCode || 200, respondToRequest);
    },

    stop: function stop() {
        nock.cleanAll();
    },

    withoutContent: function withoutContent() {
        respondWith = null;
    },

    fails: function fails() {
        respondWith = 'failure';
    },

    timeoutError: function timeoutError(options) {
        options = options || {};
        options.port = options.port || PLEX_SERVER_PORT;
        options.delay = options.delay || 3000;
        options.contentType = options.contentType || 'application/json';
        respondWith = 'content';

        return nock('http://localhost:' + options.port)
            .defaultReplyHeaders({
                'Content-Type': options.contentType,
            })
            .filteringPath(replaceActualPathToRoot)
            .get('/')
            .socketDelay(options.delay)
            .reply(options.statusCode || 200, respondToRequest);
    },
};


lib/uri.js
var uriResolvers = {
    directory: function directory(parentUrl, dir) {
        addDirectoryUriProperty(parentUrl, dir);
    },

    server: function server(parentUrl, srv) {
        addServerUriProperty(srv);
    },
};

function addServerUriProperty(server) {
    server.uri = '/system/players/' + server.address;
}

function addDirectoryUriProperty(parentUrl, directory) {
    if (parentUrl[parentUrl.length - 1] !== '/') {
        parentUrl += '/';
    }
    if (directory.key[0] === '/') {
        parentUrl = '';
    }
    directory.uri = parentUrl + directory.key;
}

exports.attach = function attach(parentUrl) {
    return function resolveAndAttachUris(result) {
        var children = result._children || [];

        children.forEach(function (child) {
            var childType = child._elementType.toLowerCase();
            var resolver = uriResolvers[childType];

            if (resolver) {
                resolver(parentUrl, child);
            }
        });

        return result;
    };
};


lib/api.js
var os = require('os');
var uuid = require('uuid');
var url = require('url');
var request = require('request');
var xml2js = require('xml2js');
var headers = require('plex-api-headers');
var extend = require('util')._extend;

var uri = require('./uri');

var PLEX_SERVER_PORT = 32400;

function PlexAPI(options, deprecatedPort) {
    var opts = options || {};
    var hostname = typeof options === 'string' ? options : options.hostname;

    this.hostname = hostname;
    this.port = deprecatedPort || opts.port || PLEX_SERVER_PORT;
    this.https = opts.https;
    this.requestOptions = opts.requestOptions || {};
    this.timeout = opts.timeout;
    this.username = opts.username;
    this.password = opts.password;
    this.managedUser = opts.managedUser;
    this.authToken = opts.token;
    this.authenticator = opts.authenticator || this._credentialsAuthenticator();
    this.responseParser = opts.responseParser || this._defaultResponseParser;
    this.options = opts.options || {};
    this.options.identifier = this.options.identifier || uuid.v4();
    this.options.product = this.options.product || 'Node.js App';
    this.options.version = this.options.version || '1.0';
    this.options.device = this.options.device || os.platform();
    this.options.deviceName = this.options.deviceName || 'Node.js App';
    this.options.platform = this.options.platform || 'Node.js';
    this.options.platformVersion = this.options.platformVersion || process.version;

    if (typeof this.hostname !== 'string') {
        throw new TypeError('Invalid Plex Server hostname');
    }
    if (typeof deprecatedPort !== 'undefined') {
        console.warn('PlexAPI constuctor port argument is deprecated, use an options object instead.');
    }

    this.serverUrl = hostname + ':' + this.port;
    this._initializeAuthenticator();
}

PlexAPI.prototype.getHostname = function getHostname() {
    return this.hostname;
};

PlexAPI.prototype.getPort = function getPort() {
    return this.port;
};

PlexAPI.prototype.getIdentifier = function getIdentifier() {
    return this.options.identifier;
};

PlexAPI.prototype.query = function query(options) {
    if (typeof options === 'string') {
        // Support old method of only supplying a single `url` parameter
        options = { uri: options };
    }
    if (options.uri === undefined) {
        throw new TypeError('Requires uri parameter');
    }

    options.method = 'GET';
    options.parseResponse = true;

    return this._request(options).then(uri.attach(options.uri));
};

PlexAPI.prototype.postQuery = function postQuery(options) {
    if (typeof options === 'string') {
        // Support old method of only supplying a single `url` parameter
        options = { uri: options };
    }
    if (options.uri === undefined) {
        throw new TypeError('Requires uri parameter');
    }

    options.method = 'POST';
    options.parseResponse = true;

    return this._request(options).then(uri.attach(url));
};

PlexAPI.prototype.putQuery = function putQuery(options) {
    if (typeof options === 'string') {
        // Support old method of only supplying a single `url` parameter
        options = { uri: options };
    }
    if (options.uri === undefined) {
        throw new TypeError('Requires uri parameter');
    }

    options.method = 'PUT';
    options.parseResponse = true;

    return this._request(options).then(uri.attach(url));
};

PlexAPI.prototype.deleteQuery = function deleteQuery(options) {
    if (typeof options === 'string') {
        // Support old method of only supplying a single `url` parameter
        options = { uri: options };
    }
    if (options.uri === undefined) {
        throw new TypeError('Requires uri parameter');
    }

    options.method = 'DELETE';
    options.parseResponse = false;

    return this._request(options);
};

PlexAPI.prototype.perform = function perform(options) {
    if (typeof options === 'string') {
        // Support old method of only supplying a single `url` parameter
        options = { uri: options };
    }
    if (options.uri === undefined) {
        throw new TypeError('Requires uri parameter');
    }

    options.method = 'GET';
    options.parseResponse = false;

    return this._request(options);
};

PlexAPI.prototype.find = function find(options, criterias) {
    if (typeof options === 'string') {
        // Support old method of only supplying a single `url` parameter
        options = { uri: options };
    }
    if (options.uri === undefined) {
        throw new TypeError('Requires uri parameter');
    }

    return this.query(options).then(function (result) {
        var children =
            result._children ||
            result.MediaContainer.Server ||
            result.MediaContainer.Directory ||
            result.MediaContainer.Provider ||
            result.MediaContainer.Metadata;
        return filterChildrenByCriterias(children, criterias);
    });
};

PlexAPI.prototype._request = function _request(options) {
    var reqUrl = this._generateRelativeUrl(options.uri);
    var method = options.method;
    var timeout = this.timeout;
    var parseResponse = options.parseResponse;
    var extraHeaders = options.extraHeaders || {};
    var self = this;

    var requestHeaders = headers(
        this,
        extend(
            {
                Accept: 'application/json',
                'X-Plex-Token': this.authToken,
                'X-Plex-Username': this.username,
            },
            extraHeaders
        )
    );

    var reqOpts = {
        uri: url.parse(reqUrl),
        encoding: null,
        method: method || 'GET',
        timeout: timeout,
        gzip: true,
        headers: requestHeaders,
        ...this.requestOptions,
    };

    return new Promise((resolve, reject) => {
        request(reqOpts, function onResponse(err, response, body) {
            if (err) {
                return reject(err);
            }

            // 403 forbidden when managed user does not have sufficient permission
            if (response.statusCode === 403) {
                return reject(
                    new Error(
                        'Plex Server denied request due to lack of managed user permissions! ' +
                            'In case of a delete request, delete content mustbe  allowed in plex-media-server options.'
                    )
                );
            }

            // 401 unauthorized when authentication is required against the requested URL
            if (response.statusCode === 401) {
                if (self.authenticator === undefined) {
                    return reject(
                        new Error(
                            'Plex Server denied request, you must provide a way to authenticate! ' +
                                'Read more about plex-api authenticators on https://www.npmjs.com/package/plex-api#authenticators'
                        )
                    );
                }

                return resolve(
                    self._authenticate().then(function () {
                        return self._request(options);
                    })
                );
            }

            if (response.statusCode < 200 || response.statusCode > 299) {
                return reject(
                    new Error(
                        'Plex Server didnt respond with a valid 2xx status code, response code: ' + response.statusCode
                    )
                );
            }

            // prevent holding an open http agent connection by pretending to consume data,
            // releasing socket back to the agent connection pool: http://nodejs.org/api/http.html#http_agent_maxsockets
            response.on('data', function onData() {});

            return parseResponse ? resolve(self.responseParser(response, body)) : resolve();
        });
    });
};

PlexAPI.prototype._authenticate = function _authenticate() {
    return new Promise((resolve, reject) => {
        if (this.authToken) {
            return reject(
                new Error(
                    'Permission denied even after attempted authentication :( Wrong username and/or password maybe?'
                )
            );
        }

        this.authenticator.authenticate(this, (err, token) => {
            if (err) {
                return reject(new Error('Authentication failed, reason: ' + err.message));
            }
            this.authToken = token;
            resolve();
        });
    });
};

PlexAPI.prototype._credentialsAuthenticator = function _credentialsAuthenticator() {
    var credentials;

    if (this.username && this.password) {
        credentials = require('plex-api-credentials');
        return credentials({
            username: this.username,
            password: this.password,
            managedUser: this.managedUser,
        });
    }
    return undefined;
};

PlexAPI.prototype._initializeAuthenticator = function _initializeAuthenticator() {
    if (this.authenticator && typeof this.authenticator.initialize === 'function') {
        this.authenticator.initialize(this);
    }
};

PlexAPI.prototype._generateRelativeUrl = function _generateRelativeUrl(relativeUrl) {
    return this._serverScheme() + this.serverUrl + relativeUrl;
};

PlexAPI.prototype._serverScheme = function _serverScheme() {
    if (typeof this.https !== 'undefined') {
        // If https is supplied by the user, always do what it says
        return this.https ? 'https://' : 'http://';
    }
    // Otherwise, use https if it's on port 443, the standard https port.
    return this.port === 443 ? 'https://' : 'http://';
};

PlexAPI.prototype._defaultResponseParser = function _defaultResponseParser(response, body) {
    if (response.headers['content-type'] === 'application/json') {
        return Promise.resolve(body.toString('utf8')).then(JSON.parse);
    } else if (!response.headers['content-type']) {
        return Promise.resolve(body.toString('utf8'));
    } else if (response.headers['content-type'].indexOf('xml') > -1) {
        return xmlToJSON(body.toString('utf8'), {
            attrkey: 'attributes',
        });
    }

    return Promise.resolve(body);
};

function xmlToJSON(str, options) {
    return new Promise((resolve, reject) => {
        xml2js.parseString(str, options, (err, jsonObj) => {
            if (err) {
                return reject(err);
            }
            resolve(jsonObj);
        });
    });
}

function filterChildrenByCriterias(children, criterias) {
    var context = {
        criterias: criterias || {},
    };

    return children.filter(criteriasMatchChild, context);
}

function criteriasMatchChild(child) {
    var criterias = this.criterias;

    return Object.keys(criterias).reduce(function matchCriteria(hasFoundMatch, currentRule) {
        var regexToMatch = new RegExp(criterias[currentRule]);
        return regexToMatch.test(child[currentRule]);
    }, true);
}

module.exports = PlexAPI;
